### ISSUE TITLE
Aggregate function assertions

### ISSUE THREAD

**Issue by 00ff0000red** [microsoft/TypeScript#42253]
# Suggestion

## 🔍 Search Terms

Multi-assertions, function assertion signatures, type assertions, aggregate assertions

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Allow functions that assert a type or condition to assert multiple types or conditions

(Maybe?) Allow those same functions to return a type other than `void`.

## 📃 Motivating Example

N/A?

## 💻 Use Cases

It allows multiple type assertions at once.
```ts
// declare AssertionError

function assert(x: unknown, y: unknown): asserts x is number & asserts y is number {
    if (typeof x !== "number" || typeof y !== "number") {
        throw new AssertionError("Assert failed: 'x' and 'y' are not both numbers");
    }
}
```

or
```
function safeAdd(x: unknown, y: unknown): number & asserts x is number & asserts y is number {
    if (typeof x !== "number" || typeof y !== "number") {
        throw new AssertionError("`x` and `y` are not both numbers");
    } else {
        return x + y;
    }
}
```

or
```ts
declare function f(...args: unknown[]): asserts args is [number, number];
```

No idea what the syntax would be, but this should give the idea.

**Comment by 00ff0000red** [comment]
Off-topic, but personally, error TS2775 is killing me.

I had a function that accepted two parameters, it only operated if the first parameter was of a type (`T`), it also had an overload for the second parameter being the same type (`T`) or anything else.

```ts
decalre function assertIsT (x: unknown): asserts x is T;
declare function isT (x: unknown): x is T;
function f (x: unknown, y: unknown): asserts x is T;
function f (x: T, y: T | H) {
    assertIsT(x);
    // x: T

    if ( isT(y) ) {
        // ... y: T
    } else {
        // ... y: H
    }
}
```
This was meant to be "JavaScript-compatible" code, as in, it could be used by a non-TS user with relative type safety, but, the assertion seems like extra work, as if I'm being over-cautious, so I wanted to simplify it:
```ts
declare function isT (x: unknown, y: unknown): asserts x is T & y is T;
function f (x: unknown, y: unknown): asserts x is T;
function f (x: T, y: T | H) {
    if ( isT(x, y) ) {
        // ... x: T
        // ... y: T
    } else {
        // ... x: T
        // ... y: H
    }
}
```

Now, it's terse and straightforward, while retaining type-safety.
(*Once again, I am not advocating any specific syntax for the aggregation/mixing*)

**Comment by 00ff0000red** [comment]
Update, just figured out that the following issues already cover the second part of what I'd have like see in TS,
> (Maybe?) Allow functions with "asserts" signatures to return a type other than void

* https://github.com/microsoft/TypeScript/issues/40562
* https://github.com/microsoft/TypeScript/issues/34636

I guess this should be scoped solely to returning multiple `x is T`, and `asserts x is T`.
And multiple `asserts x`; it wouldn't be of particular use to me as an individual, but plenty of other developers might like to have it too.

**Comment by emilioplatzer** [comment]
I also need multiple asserts guard type. Here is my Use Case:

I have a class Application (or any class that aquires resources). When is properly connected the class have a database opened connection, a path to `/tmp` and  a "log" resource. When is not connected any of than may be null. 90% of methods apply when de application is properly connected:

```ts
class Application{
    database Database | null = null
    log Log | null = null
    tmpPath string | null = null
    async connectToDb(conn:string){
         this.database = pg.connect(conn);
    }
    async setPath(path:string){
         ....
    }
    async logTo(path:string){
         ....
    }
    isProperlyConnected(database:Database|null, log:Log|null, path:string|null) asserts database is Database & log is Log & path is string{
      if(databse == null) throw new Error('database is null'); 
      ...
   }
    async doStuff1(){
        this.isProperlyConnected(this.database, this.log, this.tmpPath);
        this.database.query('select ....') ...
    }
    async doStuff2(){
        this.isProperlyConnected(this.database, this.log, this.tmpPath);
        this.log.redirectTo(this.tmpPath) ...
    }
}
```

What are the adventages?

If in the feature I need to add other resource to isProperlyConnected I do it and I warn if I forget to change it un some stuff. But if I call separate guards for each resource I do not warn if I forget one.

**Comment by Pyrolistical** [comment]
Also wrapping individual assertion functions doesn't work. The assertion doesn't seem to be passed up transitively.

```ts
function assertNumber(value: unknown): asserts value is number {
    if (typeof value !== 'number') throw new Error('expected value to be a number')
}

// this doesn't assert anything
function assertBothNumber(x: unknown, y: unknown) {
    assertNumber(x)
    assertNumber(y)
}
```
[playground demo](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAUQB4AcCm0AqBPLAHmQD4AKAfUw2ykwBMAuFASkQG8BfAWACg+ICAM5REYOAEEhQzACdYCABrNwAa3EB3JAF5EARgHDR4qTPnwwATRVh1cLYl0AmPmiy4CmQmBABbAEZy5CbScgpgiiyuNB5EPgFBZCFm4ZZR-LygkOGIAIah8gByfoGyZABuuQA2IJg2dloszPkpQoiVNZiIMG3xpRx8iEPdwIhkUJ5wox21iACE2roA5H1yS2xQABay9mKYGiiyO2VL1O509O3Vs1BwiIF5YiVr6Ty8gmAiiPQwsrSmYQsykQak0On0hk+oh+f2gAPMCGsINsYMciBcvBaYWKCTKMP+BXCkT4WKKzzxvwJKQsaWi53wcXJ5HxcMJQPSbloDK8q1kzMprOpiPSfCy0AseUJACE4FscaUyKh6mCADSIPDK+xgNjsQbDUlQeVyRXpYaSlJGsp4V6Qr4wMAsqDwomahy6AzvIzdB0Cp1sxGu8EYg0yuXksj2x3OoFqyO+6PCulczzeJkRn2wv1CiIcmJQbmp3HkOOZhNWdJAA)

Perhaps it is an easier way forward to propagated assertion functions?

**Comment by Phosra** [comment]
@Pyrolistical Ideally your example of `assertBothNumber` would have the same type as the poster's `assert(x: unknown, y: unknown): (asserts x is number) & (asserts y is number)`

**Comment by Pyrolistical** [comment]
@Phosra I agree, but perhaps this the desired behaviour can be delivered without a syntax change. That would reduce the scope.

**Comment by sebinsua** [comment]
I would also like this.

My use case is wanting to be able to narrow multiple arrays into the same opaque/branded/flavoured type when they are of equal length. This would [help us to create type-safe data operations for manipulation of arrays and matrices](https://gist.github.com/sebinsua/e756554faaa2d346aff0341374ee792e), etc.
